File: server/src/model/dto.rs
/// DTO: Data Transfer Object
///
/// All dtos should be put into this file.
use std::{fmt::Debug, ops::Deref};

use axum::{extract::Multipart, http::HeaderMap};

use serde::{Deserialize, Serialize};


use crate::toent::PossibleToent;

use super::db::{
    chnot::{ChnotKind, ChnotMetadata, ChnotRecord},
    resource::Resource,
};

#[derive(Debug, Clone, Serialize)]
pub struct KReq<E: Debug + Clone + Serialize> {
    pub body: E,
    pub namespace: String,
}

pub fn read_namespace_from_header(headers: &HeaderMap) -> String {
    headers
        .get("K-namespace")
        .and_then(|v| v.to_str().ok().map(|e| e.to_string())).unwrap()
}

pub fn kreq<E: Debug + Clone + Serialize>(headers: HeaderMap, body: E) -> KReq<E> {
    KReq {
        body,
        namespace: read_namespace_from_header(&headers),
    }
}

impl<E: Debug + Clone + Serialize> Deref for KReq<E> {
    type Target = E;

    fn deref(&self) -> &Self::Target {
        &self.body
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Chnot {
    pub record: ChnotRecord,
    pub meta: ChnotMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotUpdateReq {
    pub chnot_meta_id: String,

    pub update_time: bool,

    pub pinned: Option<bool>,
    pub archive: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotUpdateRsp {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotOverwriteReq {
    pub chnot: ChnotRecord,
    pub kind: ChnotKind,
}

impl Deref for ChnotOverwriteReq {
    type Target = ChnotRecord;

    fn deref(&self) -> &Self::Target {
        &self.chnot
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotOverwriteRsp {
    pub chnot: Chnot
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotDeletionReq {
    pub chnot_id: String,
    /// logic or physical deletion
    pub logic: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotDeletionRsp {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotQueryReq {
    pub query: Option<String>,

    // Paging
    pub start_index: u64,
    pub page_size: u64,
    pub with_deleted: Option<bool>,
    pub with_omitted: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotQueryRsp<T> {
    pub data: T,
    pub start_index: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NamespaceQueryReq {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NamespaceQueryRsp {}

pub type ResourceUploadReq = Multipart;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceUploadRsp {
    pub(crate) resources: Vec<Resource>,
}

#[derive(Clone, Debug, Deserialize)]
pub struct ToentGuessReq {
    pub input: String,
}

#[derive(Clone, Debug, Serialize)]
pub struct ToentGuessRsp {
    pub toents: Vec<PossibleToent>
}

File: web/src/store/chnot/dto.ts
import { ChnotRecord, ChnotMetadata, ChnotTag } from "./db";
import path from 'path';

export interface Chnot {
  record: ChnotRecord;
  meta: ChnotMetadata;
}

export type ListViewType = { kind: "timeline" } | { kind: "tagtree", tagkind: "children" | "descendants", tagpath: string }

export interface ChnotQueryReq {
  record_id?: string;
  meta_id?: string;
  view_type: ListViewType,

  with_omited?: boolean;
  query?: string;
  start_index: number;
  page_size: number;
}

export interface ChnotQueryRsp {
  data: Chnot[];

  start_index: number;
}

export interface ChnotOverwriteReq {
  id?: string,
  meta_id?: string,
  content: string,
  kind: string,
  insert_time: Date,
}

export interface ChnotOverwriteRsp {
  chnot: Chnot;
}

export interface ChnotDeletionReq {
  chnot_id: string;
  logic: boolean;
}

export interface ChnotUpdateReq {
  meta_id: string;

  namespace?: string;

  pinned?: boolean;
  archive?: boolean;

  update_time: boolean;
}

export interface ChnotCommentAddReq {
  id: string;

  chnot_meta_id: string;
  content: string;

  insert_time: Date;
}

export interface ChnotTagQueryReq {
  query?: string;
  query_type: "exact" | "fuzzy" | "full" | "children";

  start_index: number;
  page_size: number;
}

export interface ChnotTagQueryRsp {
  data: ChnotTag[];

  start_index: number;
}

export interface ChnotTagNamesRsp {
  data: string[];

  start_index: number;
}

export interface Toent {
  id: string;
  input: string;
  event: string;
}

export interface ToentGuessReq {
  input: string;
}

export interface ToentGuessRsp {
  toents: Toent[]
}